sql/parser/eval.go

```diff
@@ -1159,6 +1188,66 @@ func (ctx EvalContext) evalCastExpr(expr *CastExpr) (Datum, error) {
 	return DNull, fmt.Errorf("invalid cast: %s -> %s", d.Type(), expr.Type)
 }
 
+func evalComparisonEq(left, right Datum) (Datum, error) {
+	if left == DNull || right == DNull {
+		return DNull, nil
+	}
+
+	if f, ok := cmpOps[cmpArgs{EQ, reflect.TypeOf(left), reflect.TypeOf(right)}]; ok {
+		return f.fn(left, right, nil)
+	}
+
+	return DNull, fmt.Errorf("unsupported comparison operator: <%s> %s <%s>",
+		left.Type(), EQ, right.Type())
+}
+
+// foldComparisonExpr folds a given comparison operation and its datum into an
+// equivalent operation that will hit in the cmpOps map, returning this new
+// operation, along with potentially flipped operands and a "not" flag.
+func foldComparisonExpr(op ComparisonOp, dummyLeft, dummyRight Datum) (ComparisonOp, Datum, Datum, bool) {
+	switch op {
+	case NE:
+		// NE(left, right) is implemented as !EQ(left, right).
+		return EQ, dummyLeft, dummyRight, true
+	case GT:
+		// GT(left, right) is implemented as LT(right, left)
+		return LT, dummyRight, dummyLeft, false
+	case GE:
+		// GE(left, right) is implemented as LE(right, left)
+		return LE, dummyRight, dummyLeft, false
+	case NotIn:
+		// NotIn(left, right) is implemented as !IN(left, right)
+		return In, dummyLeft, dummyRight, true
+	case NotLike:
+		// NotLike(left, right) is implemented as !Like(left, right)
+		return Like, dummyLeft, dummyRight, true
+	case NotSimilarTo:
+		// NotSimilarTo(left, right) is implemented as !SimilarTo(left, right)
+		return SimilarTo, dummyLeft, dummyRight, true
+	case IsDistinctFrom:
+		// IsDistinctFrom(left, right) is implemented as !EQ(left, right)
+		//
+		// Note the special handling of NULLs and IS DISTINCT FROM above.
```

Comment:
This code about the special handling of NULLs needs to have a pointer to where that special handling occurs. It is no longer "above".